apps/finance/contracts/Finance.sol

```diff
@@ -504,132 +552,151 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
                 emit PaymentFailure(_paymentId);
-                return;
+                break;
             }
 
             // The while() predicate prevents these two from ever overflowing
-            payment.repeats += 1;
-            payed += 1;
+            payment.executions += 1;
+            paid += 1;
 
-            _makePaymentTransaction(
+            // We've already checked the remaining budget with `_canMakePayment()`
+            _unsafeMakePaymentTransaction(
                 payment.token,
                 payment.receiver,
                 payment.amount,
                 _paymentId,
-                payment.repeats,
+                payment.executions,
                 ""
             );
         }
+
+        return paid;
+    }
+
+    function _executePaymentAtLeastOnce(uint256 _paymentId) internal {
+        uint256 paid = _executePayment(_paymentId);
+        if (paid == 0) {
+            if (_nextPaymentTime(_paymentId) <= getTimestamp64()) {
+                revert(ERROR_EXECUTE_PAYMENT_NUM);
+            } else {
+                revert(ERROR_EXECUTE_PAYMENT_TIME);
+            }
+        }
     }
 
     function _makePaymentTransaction(
         address _token,
         address _receiver,
         uint256 _amount,
         uint256 _paymentId,
-        uint64 _paymentRepeatNumber,
+        uint64 _paymentExecutionNumber,
         string _reference
     )
         internal
     {
         require(_getRemainingBudget(_token) >= _amount, ERROR_REMAINING_BUDGET);
+        _unsafeMakePaymentTransaction(_token, _receiver, _amount, _paymentId, _paymentExecutionNumber, _reference);
+    }
+
+    /**
+    * @dev Unsafe version of _makePaymentTransaction that assumes you have already checked the
+    *      remaining budget
+    */
+    function _unsafeMakePaymentTransaction(
+        address _token,
+        address _receiver,
+        uint256 _amount,
+        uint256 _paymentId,
+        uint64 _paymentExecutionNumber,
+        string _reference
+    )
+        internal
+    {
         _recordTransaction(
             false,
             _token,
             _receiver,
             _amount,
             _paymentId,
-            _paymentRepeatNumber,
+            _paymentExecutionNumber,
             _reference
         );
 
         vault.transfer(_token, _receiver, _amount);
     }
 
+    function _newPeriod(uint64 _startTime) internal returns (Period storage) {
+        // There should be no way for this to overflow since each period is at least one day
+        uint64 newPeriodId = periodsLength++;
+
+        Period storage period = periods[newPeriodId];
+        period.startTime = _startTime;
+
+        // Be careful here to not overflow; if startTime + periodDuration overflows, we set endTime
+        // to MAX_UINT64 (let's assume that's the end of time for now).
+        uint64 endTime = _startTime + settings.periodDuration - 1;
+        if (endTime < _startTime) { // overflowed
```

Comment:
could be <= to state that end-time cannot be start time (logically impossible as period > 1days as per check init/setPeriodDuration)

Reply:
Not super sure about this change; the `<` check is the "safe math" check, so it might be less obvious as `<=`.

Reply:
agree, might be better to keep this as a strict overflow check.